compute/reservations/createSharedReservation.js

```diff
@@ -0,0 +1,115 @@
+   */
+  // Instantiate a reservationsClient
+  // reservationsClient = new computeLib.ReservationsClient();
+  // Instantiate a zoneOperationsClient
+  // zoneOperationsClient = new computeLib.ZoneOperationsClient();
+
+  /**
+   * TODO(developer): Update these variables before running the sample.
+   */
+  // The ID of the project where you want to reserve resources and where the instance template exists.
+  const projectId = await reservationsClient.getProjectId();
+  // The zone in which to reserve resources.
+  const zone = 'us-central1-a';
+  // The name of the reservation to create.
+  const reservationName = 'reservation-01';
+  // The number of VMs to reserve.
+  const vmsNumber = 3;
+  // The name of an existing instance template.
+  const instanceTemplateName = 'global-instance-template-name';
+  // The location of the instance template.
+  const location = 'global';
+
+  async function callCreateComputeSharedReservation() {
+    // Create reservation for 3 VMs in zone us-central1-a by specifying a instance template.
+    const specificReservation = new compute.AllocationSpecificSKUReservation({
+      count: vmsNumber,
+      sourceInstanceTemplate: `projects/${projectId}/${location}/instanceTemplates/${instanceTemplateName}`,
+    });
+
+    // Create share settings. Share reservation with one customer project.
+    const shareSettings = new compute.ShareSettings({
+      shareType: 'SPECIFIC_PROJECTS',
+      projectMap: {
+        // The IDs of projects that can consume this reservation. You can include up to 100 consumer projects.
+        // These projects must be in the same organization as the owner project.
+        // Don't include the owner project. By default, it is already allowed to consume the reservation.
+        consumerId: {
+          projectId: 'consumerId',
+        },
+      },
+    });
+
+    // Create a reservation.
+    const reservation = new compute.Reservation({
+      name: reservationName,
+      specificReservation,
+      specificReservationRequired: true,
+      shareSettings,
+    });
+
+    const [response] = await reservationsClient.insert({
+      project: projectId,
+      reservationResource: reservation,
+      zone,
+    });
+
+    let operation = response.latestResponse;
+
+    // Wait for the create reservation operation to complete.
+    while (operation.status !== 'DONE') {
+      [operation] = await zoneOperationsClient.wait({
+        operation: operation.name,
+        project: projectId,
+        zone: operation.zone.split('/').pop(),
+      });
+    }
+
+    const createdReservation = (
+      await reservationsClient.get({
+        project: projectId,
+        zone,
+        reservation: reservationName,
+      })
+    )[0];
+
+    console.log(createdReservation);
+  }
+
+  await callCreateComputeSharedReservation();
+  // [END compute_reservation_create_shared]
+}
+
+module.exports = main;
+
+// TODO(developer): Uncomment below lines before running the sample.
```

Comment:
Hi @gryczj. My comment here will apply to both modules and to the test. What really are we looking for to know if the sample did what it was intended? We are not unit testing production APIs, we simply expect that our call to an API will have the intended result. What is that? A resource ID? Not returning a failed status code or catching an exception, etc? If you were a developer using this API in production code, what would you expect -- that's the check you could demonstrate right here. The unit test probably doesn't really need mocks and all the extra complexity; isn't it enough, really, if main doesn't throw (the process doesn't exit with a non-zero exit code)? Our samples shouldn't themselves be APIs, so no need to export anything out of this module.

Reply:
I just want to add your PRs have been great! To clarify here, the goal is that a sample should fully encapsulate the demonstration of what a developer needs to know to use our APIs successfully in their own production code. How is the API intended to be used? Our tests don't need to confirm that APIs work as expected, only that the sample works (we don't want broken samples in our docs for an unexpected reason, but we do expect that service teams have already exhaustively tested their services). Hope this helps!

Reply:
@subfuzion thank you for your comments and review :) I introduced mocks here because to create shared reservation special permission are required and also another project, that we want to use as consumer and share this reservation with. I can test this sample locally with my credentials, but I have no knowledge about our test environment and projects, which I can use here as an example in tests and this sample will throw errors.
I understand your point. All new changes in the implementation were made to allow me to use mocks and fake clients in tests and on the other hand allow user to use real clients if they want to.

Reply:
@subfuzion As @gryczj wrote, the shared reservation sample would require a separate project, with a change to organization policy to allow this project to share the reservation with other projects.

I made an "executive" call for the team that writes those samples for Python, Java, Go and Node to use mocks for those samples as a kind of an exception to the rule.

My arguments:
 - We don't really do tests that span multiple projects.
 - It'd be very easy to forget about this additional project and in a year or two we'd have to rediscover the whole thing.
 - I know that Python has separate projects for every version of the language - having tests that span multiple projects that require org policy updates would make it harder to start up a new project for new version.

We acknowledge the value of proper testing and that's why mocks were picked as a middle ground between skipping tests for those samples and writing fully functional tests that would make use of multiple projects.

Let me know if I managed to convince you or not :)

Reply:
Thanks for the added context. I still don't see the need to mock console or treat the sample like its library code and have it export a function for testing, and my points about overall test simplification still remain. I'd like to surface this up to the rest of our team for discussion, so can you please address the points I've raised specifically? Thanks!

Reply:
Please feel free to ping me directly on internal chat: tonypujals@